libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -238,64 +244,158 @@ string IRGenerator::generateInitialAssignment(VariableDeclaration const& _varDec
-		unsigned paramVars = make_shared<TupleType>(constructor->functionType(false)->parameterTypes())->sizeOnStack();
+	return {generator.code(), constructorParams};
+}
 
-		Whiskers t(R"X(
-			let programSize := datasize("<object>")
-			let argSize := sub(codesize(), programSize)
+string IRGenerator::initStateVariables(ContractDefinition const& _contract)
+{
+	IRGeneratorForStatements generator{m_context, m_utils};
+	for (VariableDeclaration const* variable: _contract.stateVariables())
+		if (!variable->isConstant() && !variable->immutable())
+			generator.initializeStateVar(*variable);
+
+	return generator.code();
+}
 
-			let memoryDataOffset := <allocate>(argSize)
-			codecopy(memoryDataOffset, programSize, argSize)
+void IRGenerator::generateImplicitConstructors(ContractDefinition const& _contract)
+{
+	auto listAllParams = [&](
+		map<ContractDefinition const*, string> const& baseParams,
+		ContractDefinition const* nextContract = nullptr) -> string
+	{
+		string paramsString;
+		if (nextContract != nullptr && baseParams.find(nextContract) != baseParams.end())
+			paramsString = baseParams.at(nextContract);
+		for (auto const& param: baseParams)
+			if (param.first != nextContract)
+				paramsString += (paramsString.empty() ? "" : ", ") + param.second;
+		return paramsString;
+	};
+
+	map<ContractDefinition const*, string> baseConstructorParams;
+	for (size_t i = 0; i < _contract.annotation().linearizedBaseContracts.size(); ++i)
+	{
+		ContractDefinition const* contract = _contract.annotation().linearizedBaseContracts[i];
+		baseConstructorParams.erase(contract);
+
+		m_context.functionCollector().createFunction(implicitConstructorName(*contract), [&]() {
+			Whiskers t(R"(
+				function <functionName>(<params><comma><baseParams>) {
+					<initBaseArguments>
+					<?hasNextConstructor>
+						<nextConstructor>(<nextParams>)
+					</hasNextConstructor>
+					<initStateVariables>
+					<?hasUserDefinedConstructor>
+						<userDefinedConstructor>(<params>)
+					</hasUserDefinedConstructor>
+				}
+			)");
+			string params;
+			if (contract->constructor())
+				for (ASTPointer<VariableDeclaration> const& varDecl: contract->constructor()->parameters())
+					params += (params.empty() ? "" : ", ") + m_context.addLocalVariable(*varDecl).commaSeparatedList();
+			t("params", params);
+			string baseParamsString = listAllParams(baseConstructorParams);
+			t("baseParams", baseParamsString);
+			t("comma", !params.empty() && !baseParamsString.empty() ? ", " : "");
+			t("functionName", implicitConstructorName(*contract));
+			pair<string, map<ContractDefinition const*, string>> evaluatedArgs = evaluateConstructorArguments(*contract);
+			baseConstructorParams.insert(evaluatedArgs.second.begin(), evaluatedArgs.second.end());
+			t("initBaseArguments", evaluatedArgs.first);
+			t("hasNextConstructor", i < _contract.annotation().linearizedBaseContracts.size() - 1);
```

Comment:
Maybe move that into the `if` and use `t("hasNextConstructor", true)` and after the `if`:
```
else
    t("hasNextConstructor", false);
```